### PR TITLE
Add new SharedTree format leveraging optional-field v2

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

Exposes the option to use the new optional-field format on SharedTree.

This does not change the default format.

### CODE REVIEW

File: packages/dds/tree/src/test/testChange.ts [hunk 1]
```diff
@@ -228,7 +228,10 @@ export const TestChange = {
 	toDelta,
 	isEmpty,
 	codec,
-	codecs: makeCodecFamily([[1, codec]]),
+	codecs: makeCodecFamily([
+		[1, codec],
+		[2, codec],
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/testChange.ts:233]: Why do we need a new one of these?

Comment (Abe27342) [reply packages/dds/tree/src/test/testChange.ts:233]: `makeEditManagerCodecs` is used with test codec as a child family.

In general one consequence of not resolving versions top-down is that even when families are structured such that they don't depend on the exact logic encoding/decoding 'subdata', they still take a dependency on the versions supported for encoding that data (`makeEditManagerCodecs` depends on the fact that the codec family for its changes has versions 1 and 2). Pretty awkward IMO.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/testChange.ts:233]: Aggreed

**Review by yann-achard-MS — APPROVED** [review]

**Comment by Abe27342** [comment]
Note: I had to bump the timeout for the "all scenarios" combinatorial test because adding this additional codec means tree takes longer to initialize (due to compiling more schemas for the encoded format).

We could recover the time spent to construct a tree by doing the 'lazy codec creation' optimization that's mentioned in compatibility.md. I don't think this is much of a big deal in practice for scenarios we're targeting today (that test constructs *many* trees), and we can always optimize for it later if we want to.

**Comment by Abe27342** [comment]
Note for posterity: this PR should have updated the implementation of MessageCodec and EditManagerCodec to actually *write* `version: 2` into the persisted data as well. That's handled in #20636, but the option shouldn't have been exposed on the public API before that was sorted out.

**Pull request merged**